### PR TITLE
Support Multiple inverter api for inverter type

### PR DESCRIPTION
**Pull request by kdehairy**
The main problem is that inverters apis are different in two ways:

- The api end point to get the data
- Some aspects of parsing the non-"data" attributes of the response (like SN, version, etc...)

Each single inverter could expose (for example) two end points in general. as in, before version x it exposes an end point with query parameters, but after version x it exposes the end point with params in the request body.

The inheritance model in the project is short handed here. As it is based on the assumption that a single inverter will always support a single api end point contract.

This patch is basically (and only for now), refactoring in favor of composition (vs. inheritance).

Each inverter class is initiated with two dependencies/components:

- InverterHttpClient
- ResponseParser

To support multiple variants for each inverter, we only build the inverter with proper variations of those dependencies.

So, After this patch, the discovery process is all about asking the inverter class to build all possible variants this inverter knows it supports. The discovery then goes on, as was before, trying to use that variant to get_data, until one variant is successful.

The patch tries as much as possible to continue providing sane and most common defaults for inveter's implementations.

At this moment, the patch does not add (or subtract) any feature/support on top of the baseline (before the patch). This is just opening the opportunity to easily add missing variants.

### CODE REVIEW

**Comment by squishykid** [comment]
Thank you for the PR! I'll take some time to review it. In the mean time, it should now be possible for you to run the CI tooling.

My first comment is that there seem to be a lot of whitespace changes, which `black` is also complaining about. I'll swing back once the build is green!

**Comment by kdehairy** [comment]
thnx. I amended my commit with the expected formatting. But I'm still not allowed to run the CI checks. Can you please give it a go for me?

**Comment by squishykid** [comment]
I think I will have to approve each run until you get your first contribution in to the repo. After that it should run automatically

**Comment by kdehairy** [comment]
Alright, I ran all the checks locally and fixed (to my best at least). Hopefully this next run will pass. Can you give it a go again? thnx

**Comment by squishykid** [comment]
You should run the `black` code formatter locally before pushing, that will also reduce the size of the diff

**Comment by kdehairy** [comment]
I did. But then I modified code to address the reset of the checks, and skipped my mind to run it again :) .. sorry

I fixed. And added a `verify.sh` script to run the checks locally. Might help next "first-contributor" to avoid back and forth communication with maintainers to run the checks manually for them.

**Comment by squishykid** [comment]
I've been thinking about adding something like that. Thanks!

**Comment by kdehairy** [comment]
Did you have a chance to review it? :)

File: solax/__init__.py [hunk 1]
```diff
@@ -1,6 +1,5 @@
 """Support for Solax inverter via local API."""
 import asyncio
-
```
Comment (squishykid) [inline solax/__init__.py:3]: From an initial glance, I can see a few things we will need to change:
- There are a lot of whitespace and non-functional changes. I am referring to changes where you have reordered imports or removed whitespace between imports. These pollute the diff an will have to be removed
- I notice that you added a few logger statements. As this is a library, we should not be logging anything and leave the decision on how to bubble-up errors and information to the program calling the library, so you will have to remove the loggers too.

File: solax/discovery.py [hunk 2]
```diff
@@ -26,6 +28,9 @@
     X1HybridGen4,
 ]
 
+_LOGGER = logging.getLogger(__name__)
```
Comment (squishykid) [inline solax/discovery.py:31]: These are one instance of the loggers that will have to be removed

File: tests/test_solax.py [hunk 3]
```diff
@@ -1,6 +1,8 @@
-from unittest.mock import Mock
 import asyncio
```
Comment (squishykid) [inline tests/test_solax.py:1]: It seems like there are no functional changes here, just reordering the imports. As mentioned above, this pollutes the diff.

File: tests/test_vol.py [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import pytest
 from voluptuous import Invalid
+
```
Comment (squishykid) [inline tests/test_vol.py:3]: Whitespace change that pollutes the diff

File: verify.sh [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+mypy --exclude venv .
+
+echo "Running flake8..."
+flake8 --ignore=E501 solax tests
+
+echo "Running pylint..."
+pylint -d 'C0111' solax tests
+
+echo "Running pytest..."
+pytest --cov=solax --cov-fail-under=100 --cov-branch --cov-report=term-missing .
```
Comment (squishykid) [inline verify.sh:23]: Thank you for adding this file!

**Review by squishykid — CHANGES_REQUESTED** [review]

**Comment by kdehairy** [comment]
Thanks for putting the time. I addressed the comments.

Although, I did not find a unified import sorting/ordering style in the code base (or any tool checking it). 

Do you find it a good idea to include something like `isort`?

File: solax/inverters/__init__.py [hunk 6]
```diff
@@ -1,13 +1,13 @@
...
 from .x_hybrid import XHybrid
 from .x1 import X1
+from .x1_boost import X1Boost
 from .x1_hybrid_gen4 import X1HybridGen4
 from .x1_mini import X1Mini
 from .x1_mini_v34 import X1MiniV34
 from .x1_smart import X1Smart
 from .x3_v34 import X3V34
 from .x3 import X3
-from .x1_boost import X1Boost
```
Comment (squishykid) [inline solax/inverters/__init__.py:10]: There’s still an import re-shuffle here which is unnecessary

**Comment by kdehairy** [comment]
Sorry I missed that. Addressed

**Comment by kdehairy** [comment]
Any chance to have a look?

**Comment by kdehairy** [comment]
Still no luck finding time for this PR?

**Comment by squishykid** [comment]
Ok, lets do it!

**Comment by squishykid** [comment]
Ok, i will need to have another quick check tomorrow, but I think we can merge this. Thank youuuuuu for your patience

**Comment by kdehairy** [comment]
Thnx and non worries.

**Comment by squishykid** [comment]
So my only issue here is how the inverters are constructed for the smoke tests, because it creates a list of them and then doesn’t exactly guarantee that it will test all of them.

I can fix that in a follow up PR though, so I’ll merge this and then pursue that in my own time. Thank you for this contribution, I really like the design you went with. And thanks for your patience!

**Comment by kdehairy** [comment]
Thank you for putting the time reviewing.

**Pull request merged**